README.md
# NAME

Perl5::CoreSmokeDB::Schema - [DBIC::Schema](https://metacpan.org/pod/DBIx::Class::Schema) for the smoke reports database

# SYNOPSIS

```perl
use Perl5::CoreSmokeDB::Schema;
my $schema = Perl5::CoreSmokeDB::Schema->connect($dsn, $user, $pswd, $options);

my $report = $schema->resultset('Report')->find({ id => 1 });
```
# DESCRIPTION

This class is used in the backend for accessing the database.

Another use is: `$schema->deploy()`

# SCHEMA

This ORM is generated by
[DBIx::Class::Schema::Loader](https://metacpan.org/pod/DBIx::Class::Schema::Loader)
and the `dbicdump` tool it provides (also see
[dbicdump.conf](support/dbicdump.conf)).

Central to the schema are `report`s, they have `config`s. Each `config` has `result`s, and each `result` may have `failures_for_env` that point to a `failure`.

![Perl5::CoreSmokeDB::Schema](support/schema.png)

## SmokeConfig

At the moment this table is not actively used (as the data is not propagated by
[Test::Smoke](https://metacpan.org/pod/Test::Smoke)), but each `report` belongs
to a `smoke_config`.

## Report

This table holds the basic information for a Perl5 core smoke report that is consistent throughout the smoke-run (like OS, hardware, source-tree status and aggregate information).

## Config

A `config` consists the arguments passed to `./Configure` the name and version
of the c-compiler.

## Result

This is the aggregate information of a single `make test` run, recording the
value of the `PERLIO` environment variable the summary (`P`/`F`) and the timing
aggregates.

## FailuresForEnv

This is an intermediate table to normalise the many-to-many relationship
between a `result` and a test-`failure`.

## Failure

This table holds the test name, status and output from the test harness.

## TsgatewayConfig

This table (`tsgateway_config`) holds the schema version of the database and
was only introduced after the first version of the schema. It might also be
used future to keep other global information about the `Perl5::CoreSmokeDB`
software.

# SEE ALSO

* [Perl5::CoreSmokeDB::API](https://github.com/abeltje/Perl5-CoreSmokeDB-API#name)
* [Perl5::CoreSmokeDB::Web](https://github.com/abeltje/Perl5-CoreSmokeDB-Web#name)

# AUTHOR

&copy; MMXIII - MMXXII Abe Timmerman <abeltje@cpan.org>, H.Merijn Brand

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.



Changes
1.12 2023-06-22T13:47:28+02:00 (ebbbce8 => Abe Timmerman)
 - (Abe Timmerman, Thu, 22 Jun 2023 13:47:28 +0200) Autocommit for
   distribution Perl5-CoreSmokeDB-Schema 1.12 (minor)

1.11 2023-05-15T14:06:59+02:00 (12ac6b3 => Abe Timmerman)
 - (Abe Timmerman, Mon, 15 May 2023 14:06:59 +0200) Autocommit for
   distribution Perl5::CoreSmokeDB::Schema 1.11 (minor)
 - (Abe Timmerman, Thu, 22 Jun 2023 13:42:44 +0200) Define
   InflateColumn::Bytea
 -     When retrieving bytea fields from the database, we should decode
   them
 -     from utf-8 to characters. The way to the database is more
   complicated:
 -     Depending on the way this report data is posted to the service we
   need
 -     to add an extra encode() for those bytea fields. That is not done
   here.

1.10_01 2023-05-15T13:44:30+02:00 (262cf91 => Abe Timmerman)
 - (Abe Timmerman, Mon, 15 May 2023 13:44:30 +0200) Autocommit for
   distribution Perl5::CoreSmokeDB::Schema 1.10_01 (test)

1.10 2023-05-02T12:13:05+02:00 (8295089 => Abe Timmerman)
 - (Abe Timmerman, Tue, 2 May 2023 12:13:05 +0200) Autocommit for
   distribution Perl5::CoreSmokeDB::Schema 1.10 (minor)
 - (Abe Timmerman, Tue, 2 May 2023 15:22:59 +0200) Fix the copyright year
   range...
 -     14:48:18    @[Tux] | abeltje,
 -	 https://metacpan.org/pod/Perl5::CoreSmokeDB::Schema "MMXIII-
   MMXII"
 -	 "MMXII - MMXXIII". De eerste reekst is al vreemd, maar
 -	 ik ben niet op de hoogte van bewezen feiten dat de Romeinen
 -	 achteruit konden tellen
 - (Abe Timmerman, Mon, 15 May 2023 13:41:21 +0200) Fix: failures weren't
   expanded for full_report_data
 -     See https://github.com/abeltje/Perl5-CoreSmokeDB-API/issues/1

1.09_01 2023-05-01T16:17:06+02:00 (7e76b37 => Abe Timmerman)
 - (Abe Timmerman, Mon, 1 May 2023 16:17:06 +0200) Autocommit for
   distribution Perl5::CoreSmokeDB::Schema 1.09_01 (test)

1.09 2023-04-27T22:42:31+02:00 (9833269 => Abe Timmerman)
 - (Abe Timmerman, Thu, 27 Apr 2023 22:42:31 +0200) Autocommit for
   distribution Perl5::CoreSmokeDB::Schema 1.09 (minor)
 - (Abe Timmerman, Mon, 1 May 2023 16:08:12 +0200) Bugfix: we don't want
   "empty" rows in the matrix
 -     We have smokers with multiple CC incantations and different -Dxxx
 -     parameters, we only want the those configs with test results, no
   "holes"
 -     in the matrix.

1.08_04 2023-04-27T22:40:33+02:00 (a346490 => Abe Timmerman)
 - (Abe Timmerman, Thu, 27 Apr 2023 22:40:33 +0200) Autocommit for
   distribution Perl5::CoreSmokeDB::Schema 1.08_04 (same)

1.08_03 2022-10-19T17:26:19+02:00 (f7d802a => Abe Timmerman)
 - (Abe Timmerman, Wed, 19 Oct 2022 17:26:19 +0200) Autocommit for
   distribution Perl5::CoreSmokeDB::Schema 1.08_03 (same)
 - (Abe Timmerman, Thu, 20 Oct 2022 12:40:21 +0200) Add description for the
   TsgatewayConfig table
 - (Abe Timmerman, Thu, 20 Oct 2022 13:12:07 +0200) Fix typo

1.08 2022-10-12T13:20:56+02:00 (febb462 => Abe Timmerman)
 - (Abe Timmerman, Wed, 12 Oct 2022 13:20:56 +0200) Autocommit for
   distribution Perl5::CoreSmokeDB::Schema 1.08 (minor)
 - (Abe Timmerman, Wed, 19 Oct 2022 12:47:41 +0200) Update dependencies
 -     We need DateTime 1.53 for the `rfc3339` method.
 - (Abe Timmerman, Wed, 19 Oct 2022 13:03:11 +0200) Autocommit for
   distribution Perl5::CoreSmokeDB::Schema 1.0801 (test)
 - (Abe Timmerman, Wed, 19 Oct 2022 15:55:39 +0200) Add more information
   about the Schema
 -     Adding description of the tables and picture in README.md
 -     The picture is generated from the ERD-tool in pgAdmin4
 - (Abe Timmerman, Wed, 19 Oct 2022 15:59:16 +0200) Autocommit for
   distribution Perl5::CoreSmokeDB::Schema 1.0802 (test)
 - (Abe Timmerman, Wed, 19 Oct 2022 17:22:41 +0200) The version must be a
   string if one wants Beta versions

1.07 2022-09-25T19:58:32+02:00 (d58714f => Abe Timmerman)
 - (Abe Timmerman, Sun, 25 Sep 2022 19:58:32 +0200) Autocommit for
   distribution Perl5::CoreSmokeDB::Schema 1.07 (same)
 - (Abe Timmerman, Wed, 12 Oct 2022 13:11:17 +0200) Fix warning in result
   classes
 -     The classes with ->as_hashref as a method warned for the simple case
 -     that '$is_full' was not passed.

1.06 2022-09-25T18:41:54+02:00 (c1f5937 => Abe Timmerman)
 - (Abe Timmerman, Sun, 25 Sep 2022 18:41:54 +0200) Autocommit for
   distribution Perl5::CoreSmokeDB::Schema 1.06 (same)
 - (Abe Timmerman, Sun, 25 Sep 2022 19:54:34 +0200) Rename the
   Exception-classes the Schema throws
 -     The Exception::Class exception were still called
 -     'Test::Smoke::Gateway::*Exception', now that we've pulled the schema
   out
 -     of the CoreSmokeDB (Test::Smoke::Gateway) repository, better to
   rename
 -     the Exception classes to 'Perl5::CoreSmokeDB::Schea::*Exception'

t/lib/APIish.pm
package APIish;
use warnings;
use strict;

use Exporter 'import';
our @EXPORT = qw< read_json get_report_data create_full_report >;

use Date::Parse qw( str2time );
use DateTime;
use JSON;

sub read_json {
    my ($filename) = @_;

    if (open(my $fh, '<', $filename)) {
        my $json = do { local $/; <$fh> };
        return from_json($json);
    }
    die "Cannot open($filename): $!";
}

sub get_report_data {
    my ($smoke_data) = @_;

    my $report_data = {
        %{ $smoke_data->{'sysinfo'} },
        sconfig_id => 0,
    };
    $report_data->{lc($_)} = delete $report_data->{$_} for keys %$report_data;
    $report_data->{smoke_date} = DateTime->from_epoch(
        epoch     => str2time($report_data->{smoke_date}),
        time_zone => 'UTC',
    );

    my @to_unarray = qw/
        skipped_tests applied_patches
        compiler_msgs manifest_msgs nonfatal_msgs
    /;
    $report_data->{$_} = join("\n", @{$smoke_data->{$_} || []}) for @to_unarray;

    my @other_data = qw/harness_only harness3opts summary/;
    $report_data->{$_} = $smoke_data->{$_} for @other_data;

    return $report_data;
}

sub create_full_report {
    my ($schema, $smoke_data) = @_;

    my $report_data = get_report_data($smoke_data);
    my $report = $schema->txn_do(
        sub {
            my $r = $schema->resultset('Report')->create($report_data);
            $r->discard_changes;

            for my $config (@{ $smoke_data->{configs}}) {
                my $results = delete($config->{'results'});
                for my $field (qw/cc ccversion/) {
                    $config->{$field} ||= '?';
                }

                $config->{started} = DateTime->from_epoch(
                    epoch     => str2time($config->{started}),
                    time_zone => 'UTC',
                );

                my $conf = $r->create_related('configs', $config);

                for my $result (@$results) {
                    my $failures = delete($result->{'failures'});
                    my $res = $conf->create_related('results', $result);

                    for my $failure (@$failures) {
                        $failure->{'extra'} = join("\n", @{$failure->{'extra'}});
                        my $db_failure = $schema->resultset(
                            'Failure'
                        )->find_or_create(
                            $failure,
                            {key => 'failure_test_status_extra_key'}
                        );
                        $schema->resultset('FailureForEnv')->create(
                            {
                                result_id  => $res->id,
                                failure_id => $db_failure->id,
                            }
                        );
                    }
                }
            }

            return $r;
        }
    );
}
1;


Makefile.PL
#! perl -w
use warnings;
use strict;

use ExtUtils::MakeMaker 6.55_02;

WriteMakefile(
    NAME          => 'Perl5::CoreSmokeDB::Schema',
    VERSION_FROM  => 'lib/Perl5/CoreSmokeDB/Schema.pm',
    ABSTRACT_FROM => 'lib/Perl5/CoreSmokeDB/Schema.pm',
    AUTHOR        => 'Abe Timmerman <abeltje@cpan.org>',
    LICENSE       => 'perl',

    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 6.55_02,
    },
    BUILD_REQUIRES => {
        'Test::Fatal'        => '0.01',
        'Test::Simple'       => '0.88',
        'Test::Warnings'     => '0.028',
        'Test::DBIC::SQLite' => '1.01',
        'JSON'               => '4.09',
    },
    PREREQ_PM => {
        'DBIx::Class'                 => '0.08271',
        'DBIx::Class::Schema::Loader' => '0.07049',
        'DBD::Pg'                     => '3.14.0',
        'DateTime::Format::Pg'        => '0.16013',
    },

    MIN_PERL_VERSION => '5.010001',

    META_ADD => {
        'meta-spec' => {
            version    => 2,
        },
        'resources' => {
            homepage  => 'https://github.com/abeltje/Perl5-CoreSmokeDB-Schema.git',
            repository => {
                type   => 'git',
                url    => 'https://github.com:abeltje/Perl5-CoreSmokeDB-Schema.git',
                web    => 'https://github.com/abeltje/Perl5-CoreSmokeDB-Schema.git',
            },
#            x_IRC      => 'irc://irc.perl.org/#smoke',
        },
    },
);


lib/Perl5/CoreSmokeDB/Schema/Result/Failure.pm
use utf8;
package Perl5::CoreSmokeDB::Schema::Result::Failure;

# Created by DBIx::Class::Schema::Loader
# DO NOT MODIFY THE FIRST PART OF THIS FILE

=head1 NAME

Perl5::CoreSmokeDB::Schema::Result::Failure

=cut

use strict;
use warnings;

use base 'DBIx::Class::Core';

=head1 COMPONENTS LOADED

=over 4

=item * L<DBIx::Class::InflateColumn::DateTime>

=back

=cut

__PACKAGE__->load_components("InflateColumn::DateTime");

=head1 TABLE: C<failure>

=cut

__PACKAGE__->table("failure");

=head1 ACCESSORS

=head2 id

  data_type: 'integer'
  is_auto_increment: 1
  is_nullable: 0
  sequence: 'failure_id_seq'

=head2 test

  data_type: 'text'
  is_nullable: 0
  original: {data_type => "varchar"}

=head2 status

  data_type: 'text'
  is_nullable: 0
  original: {data_type => "varchar"}

=head2 extra

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=cut

__PACKAGE__->add_columns(
  "id",
  {
    data_type         => "integer",
    is_auto_increment => 1,
    is_nullable       => 0,
    sequence          => "failure_id_seq",
  },
  "test",
  {
    data_type   => "text",
    is_nullable => 0,
    original    => { data_type => "varchar" },
  },
  "status",
  {
    data_type   => "text",
    is_nullable => 0,
    original    => { data_type => "varchar" },
  },
  "extra",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
);

=head1 PRIMARY KEY

=over 4

=item * L</id>

=back

=cut

__PACKAGE__->set_primary_key("id");

=head1 UNIQUE CONSTRAINTS

=head2 C<failure_test_status_extra_key>

=over 4

=item * L</test>

=item * L</status>

=item * L</extra>

=back

=cut

__PACKAGE__->add_unique_constraint("failure_test_status_extra_key", ["test", "status", "extra"]);

=head1 RELATIONS

=head2 failures_for_env

Type: has_many

Related object: L<Perl5::CoreSmokeDB::Schema::Result::FailureForEnv>

=cut

__PACKAGE__->has_many(
  "failures_for_env",
  "Perl5::CoreSmokeDB::Schema::Result::FailureForEnv",
  { "foreign.failure_id" => "self.id" },
  { cascade_copy => 0, cascade_delete => 0 },
);


# Created by DBIx::Class::Schema::Loader v0.07049 @ 2022-09-06 09:15:22
# DO NOT MODIFY THIS OR ANYTHING ABOVE! md5sum:47DdjPyTSksgr5Z6Dw+HgA

=head2 $failure->as_hashref()

Returns a hashref with inflated columns

=cut

sub as_hashref {
    my $self = shift;

    my $record = { $self->get_inflated_columns };

    return $record;
}

1;


lib/Perl5/CoreSmokeDB/Schema/Result/Report.pm
use utf8;
package Perl5::CoreSmokeDB::Schema::Result::Report;

# Created by DBIx::Class::Schema::Loader
# DO NOT MODIFY THE FIRST PART OF THIS FILE

=head1 NAME

Perl5::CoreSmokeDB::Schema::Result::Report

=cut

use strict;
use warnings;

use base 'DBIx::Class::Core';

=head1 COMPONENTS LOADED

=over 4

=item * L<DBIx::Class::InflateColumn::DateTime>

=back

=cut

__PACKAGE__->load_components("InflateColumn::DateTime");

=head1 TABLE: C<report>

=cut

__PACKAGE__->table("report");

=head1 ACCESSORS

=head2 id

  data_type: 'integer'
  is_auto_increment: 1
  is_nullable: 0
  sequence: 'report_id_seq'

=head2 sconfig_id

  data_type: 'integer'
  is_foreign_key: 1
  is_nullable: 1

=head2 duration

  data_type: 'integer'
  is_nullable: 1

=head2 config_count

  data_type: 'integer'
  is_nullable: 1

=head2 reporter

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 reporter_version

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 smoke_perl

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 smoke_revision

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 smoke_version

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 smoker_version

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 smoke_date

  data_type: 'timestamp with time zone'
  is_nullable: 0

=head2 perl_id

  data_type: 'text'
  is_nullable: 0
  original: {data_type => "varchar"}

=head2 git_id

  data_type: 'text'
  is_nullable: 0
  original: {data_type => "varchar"}

=head2 git_describe

  data_type: 'text'
  is_nullable: 0
  original: {data_type => "varchar"}

=head2 applied_patches

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 hostname

  data_type: 'text'
  is_nullable: 0
  original: {data_type => "varchar"}

=head2 architecture

  data_type: 'text'
  is_nullable: 0
  original: {data_type => "varchar"}

=head2 osname

  data_type: 'text'
  is_nullable: 0
  original: {data_type => "varchar"}

=head2 osversion

  data_type: 'text'
  is_nullable: 0
  original: {data_type => "varchar"}

=head2 cpu_count

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 cpu_description

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 username

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 test_jobs

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 lc_all

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 lang

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 user_note

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 manifest_msgs

  data_type: 'bytea'
  is_nullable: 1

=head2 compiler_msgs

  data_type: 'bytea'
  is_nullable: 1

=head2 skipped_tests

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 log_file

  data_type: 'bytea'
  is_nullable: 1

=head2 out_file

  data_type: 'bytea'
  is_nullable: 1

=head2 harness_only

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 harness3opts

  data_type: 'text'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 summary

  data_type: 'text'
  is_nullable: 0
  original: {data_type => "varchar"}

=head2 smoke_branch

  data_type: 'text'
  default_value: 'blead'
  is_nullable: 1
  original: {data_type => "varchar"}

=head2 nonfatal_msgs

  data_type: 'bytea'
  is_nullable: 1

=head2 plevel

  data_type: 'text'
  default_value: git_describe_as_plevel(git_describe)
  is_nullable: 1
  original: {data_type => "varchar"}

=cut

__PACKAGE__->add_columns(
  "id",
  {
    data_type         => "integer",
    is_auto_increment => 1,
    is_nullable       => 0,
    sequence          => "report_id_seq",
  },
  "sconfig_id",
  { data_type => "integer", is_foreign_key => 1, is_nullable => 1 },
  "duration",
  { data_type => "integer", is_nullable => 1 },
  "config_count",
  { data_type => "integer", is_nullable => 1 },
  "reporter",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "reporter_version",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "smoke_perl",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "smoke_revision",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "smoke_version",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "smoker_version",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "smoke_date",
  { data_type => "timestamp with time zone", is_nullable => 0 },
  "perl_id",
  {
    data_type   => "text",
    is_nullable => 0,
    original    => { data_type => "varchar" },
  },
  "git_id",
  {
    data_type   => "text",
    is_nullable => 0,
    original    => { data_type => "varchar" },
  },
  "git_describe",
  {
    data_type   => "text",
    is_nullable => 0,
    original    => { data_type => "varchar" },
  },
  "applied_patches",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "hostname",
  {
    data_type   => "text",
    is_nullable => 0,
    original    => { data_type => "varchar" },
  },
  "architecture",
  {
    data_type   => "text",
    is_nullable => 0,
    original    => { data_type => "varchar" },
  },
  "osname",
  {
    data_type   => "text",
    is_nullable => 0,
    original    => { data_type => "varchar" },
  },
  "osversion",
  {
    data_type   => "text",
    is_nullable => 0,
    original    => { data_type => "varchar" },
  },
  "cpu_count",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "cpu_description",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "username",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "test_jobs",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "lc_all",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "lang",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "user_note",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "manifest_msgs",
  { data_type => "bytea", is_nullable => 1 },
  "compiler_msgs",
  { data_type => "bytea", is_nullable => 1 },
  "skipped_tests",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "log_file",
  { data_type => "bytea", is_nullable => 1 },
  "out_file",
  { data_type => "bytea", is_nullable => 1 },
  "harness_only",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "harness3opts",
  {
    data_type   => "text",
    is_nullable => 1,
    original    => { data_type => "varchar" },
  },
  "summary",
  {
    data_type   => "text",
    is_nullable => 0,
    original    => { data_type => "varchar" },
  },
  "smoke_branch",
  {
    data_type     => "text",
    default_value => "blead",
    is_nullable   => 1,
    original      => { data_type => "varchar" },
  },
  "nonfatal_msgs",
  { data_type => "bytea", is_nullable => 1 },
  "plevel",
  {
    data_type     => "text",
    default_value => \"git_describe_as_plevel(git_describe)",
    is_nullable   => 1,
    original      => { data_type => "varchar" },
  },
);

=head1 PRIMARY KEY

=over 4

=item * L</id>

=back

=cut

__PACKAGE__->set_primary_key("id");

=head1 UNIQUE CONSTRAINTS

=head2 C<report_git_id_smoke_date_duration_hostname_architecture_key>

=over 4

=item * L</git_id>

=item * L</smoke_date>

=item * L</duration>

=item * L</hostname>

=item * L</architecture>

=back

=cut

__PACKAGE__->add_unique_constraint(
  "report_git_id_smoke_date_duration_hostname_architecture_key",
  ["git_id", "smoke_date", "duration", "hostname", "architecture"],
);

=head1 RELATIONS

=head2 configs

Type: has_many

Related object: L<Perl5::CoreSmokeDB::Schema::Result::Config>

=cut

__PACKAGE__->has_many(
  "configs",
  "Perl5::CoreSmokeDB::Schema::Result::Config",
  { "foreign.report_id" => "self.id" },
  { cascade_copy => 0, cascade_delete => 0 },
);

=head2 sconfig

Type: belongs_to

Related object: L<Perl5::CoreSmokeDB::Schema::Result::SmokeConfig>

=cut

__PACKAGE__->belongs_to(
  "sconfig",
  "Perl5::CoreSmokeDB::Schema::Result::SmokeConfig",
  { id => "sconfig_id" },
  {
    is_deferrable => 0,
    join_type     => "LEFT",
    on_delete     => "NO ACTION",
    on_update     => "NO ACTION",
  },
);


# Created by DBIx::Class::Schema::Loader v0.07049 @ 2022-09-06 09:15:22
# DO NOT MODIFY THIS OR ANYTHING ABOVE! md5sum:YSwjDKOFBQjshqCpkN9sCA

##### Problem with DBIx::Class/DBIx::Class::Schema::Loader
# it cannot ADD COLUMN plevel varchar GENERATED ALWAYS AS (...) STORED
# so remove the default_value for the ORM and put it back with SQL in
# Test::Smoke::Gateway::Schema::deploy()
#
my $_plevel_column = __PACKAGE__->column_info('plevel');
__PACKAGE__->remove_column('plevel');
delete($_plevel_column->{default_value});
__PACKAGE__->add_column('plevel', $_plevel_column);

##### Create InflateColumn::Bytea
# we need to make sure the data is a byte-string to store it as `bytea`
# as it turns out, 'deflate' is called too late, DBIx::Class has already blown
# up with 'wide-character'. This only a problem via the old POST /report
# the new RESTish POST /api/report is fine.
# So we only define inflate, that makes sure we have a char-string.
use Encode qw< encode decode >;
my @_bytea = qw< compiler_msgs manifest_msgs nonfatal_msgs log_file out_file >;
for my $field (@_bytea) {
    __PACKAGE__->inflate_column(
        $field => {
            inflate => sub { decode('UTF8', $_[0]); }, # from database
#            deflate => sub { encode('UTF8', $_[0]); }, # to database, too late
        }
    );
}

sub arch_os_version_key {
    my $self = shift;
    return join( "##", $self->architecture, $self->osname, $self->osversion, $self->hostname);
}

sub arch_os_version_label {
    my $self = shift;
    return join( " - ", $self->architecture, $self->osname, $self->osversion, $self->hostname);
}

sub arch_os_version_pair {
    my $self = shift;
    return {value => $self->arch_os_version_key, label => $self->arch_os_version_label};
}

my %io_env_order_map = (
    minitest => 1,
    stdio    => 2,
    perlio   => 3,
    locale   => 4,
);
my $max_io_envs = scalar(keys %io_env_order_map);

sub title {
    my $self = shift;

    return join(
        " ",
        "Smoke",
        $self->git_describe,
        $self->summary,
        $self->osname,
        $self->osversion,
        $self->cpu_description,
        $self->cpu_count
    );
}

sub list_title {
    my $self = shift;

    return join(
        " ",
        $self->git_describe,
        $self->osname,
        $self->osversion,
        $self->cpu_description,
        $self->cpu_count
    );
}

sub c_compilers {
    my $self = shift;

    my %c_compiler_seen;
    my $i = 1;
    for my $config ($self->configs) {
        $c_compiler_seen{$config->c_compiler_key} //= {
            index     => $i++,
            key       => $config->c_compiler_key,
            cc        => $config->cc,
            ccversion => $config->ccversion,
        };
    }
    return [
        sort {
            $a->{index} <=> $b->{index}
        } values %c_compiler_seen
    ];
}

sub matrix {
    my $self = shift;

    my %c_compilers = map {
        $_->{key} => $_
    } @{$self->c_compilers};

    my (%matrix, %cfg_order, %io_env_seen);
    my $o = 0;
    for my $config ($self->configs) {
        for my $result ($config->results) {
            my $cc_index = $c_compilers{$config->c_compiler_key}{index};

            $matrix{$cc_index}{$config->debugging}{$config->arguments}{$result->io_env} =
                $result->summary;
            $io_env_seen{$result->io_env} = $result->locale;
        }
        $cfg_order{$config->arguments} //= $o++;
    }

    my @io_env_in_order = sort {
        $io_env_order_map{$a} <=> $io_env_order_map{$b}
    } keys %io_env_seen;

    my @cfg_in_order = sort {
        $cfg_order{$a} <=> $cfg_order{$b}
    } keys %cfg_order;

    my @matrix;
    for my $cc (sort { $a->{index} <=> $b->{index} } values %c_compilers) {
        my $cc_index = $cc->{index};
        for my $cfg (@cfg_in_order) {
            next if !exists($matrix{$cc_index}{N}{$cfg})
                 && !exists($matrix{$cc_index}{D}{$cfg});
            my @line;
            for my $debugging (qw/ N D /) {
                for my $io_env (@io_env_in_order) {
                    push(
                        @line,
                        $matrix{$cc_index}{$debugging}{$cfg}{$io_env} || '-'
                    );
                }
            }
            while (@line < 8) { push @line, " " }
            my $mline = join("  ", @line);
            push @matrix, "$mline  $cfg (\*$cc_index)";
        }
    }
    my @legend = $self->matrix_legend(
        [
            map { $io_env_seen{$_} ? "$_:$io_env_seen{$_}" : $_ }
                @io_env_in_order
        ]
    );
    return @matrix, @legend;
}

sub matrix_legend {
    my $self = shift;
    my ($io_envs) = @_;

    my @legend = (
        (map "$_ DEBUGGING", reverse @$io_envs),
        (reverse @$io_envs)
    );
    my $first_line = join("  ", ("|") x @legend);

    my $length = (3 * 2 * $max_io_envs) - 2;
    for my $i (0 .. $#legend) {
        my $bar_count = scalar(@legend) - $i;
        my $prefix = join("  ", ("|") x $bar_count);
        $prefix =~ s/(.*)\|$/$1+/;
        my $dash_count = $length - length($prefix);
        $prefix .= "-" x $dash_count;
        $legend[$i] = "$prefix  $legend[$i]"
    }
    unshift @legend, $first_line;
    return @legend;
}

sub test_failures {
    my $self = shift;
    return $self->group_tests_by_status('FAILED');
}

sub test_todo_passed {
    my $self = shift;
    return $self->group_tests_by_status('PASSED');
}

sub group_tests_by_status {
    my $self = shift;
    my ($group_status) = @_;

    use Data::Dumper; $Data::Dumper::Indent = 1; $Data::Dumper::Sortkeys = 1;

    my %c_compilers = map {
        $_->{key} => $_
    } @{$self->c_compilers};

    my (%tests);
    my $max_name_length = 0;
    for my $config ($self->configs) {
        for my $result ($config->results) {
            for my $io_env ($result->failures_for_env) {
                for my $test ($io_env->failure) {
                    next if $test->status ne $group_status;

                    $max_name_length = length($test->test)
                        if length($test->test) > $max_name_length;

                    my $key = $test->test . $test->extra;
                    push(
                        @{$tests{$key}{$config->full_arguments}{test}}, {
                            test_env => $result->test_env,
                            test     => { $test->get_inflated_columns },
                        }
                    );
                }
            }
        }
    }
    my @grouped_tests;
    for my $group (values %tests) {
        push @grouped_tests, {test => undef, configs => [ ]};
        for my $cfg (keys %$group) {
            push @{ $grouped_tests[-1]->{configs} }, {
                arguments => $cfg,
                io_envs   => join("/", map $_->{test_env}, @{ $group->{$cfg}{test} })
            };
            $grouped_tests[-1]{test} //= $group->{$cfg}{test}[0]{test};
        }
    }
    return \@grouped_tests;
}

sub duration_in_hhmm {
    my $self = shift;
    return time_in_hhmm($self->duration);
}

sub average_in_hhmm {
    my $self = shift;
    return time_in_hhmm($self->duration/$self->config_count);
}

sub time_in_hhmm {
    my $diff = shift;

    # Only show decimal point for diffs < 5 minutes
    my $digits = $diff =~ /\./ ? $diff < 5*60 ? 3 : 0 : 0;
    my $days = int( $diff / (24*60*60) );
    $diff -= 24*60*60 * $days;
    my $hour = int( $diff / (60*60) );
    $diff -= 60*60 * $hour;
    my $mins = int( $diff / 60 );
    $diff -=  60 * $mins;
    $diff = sprintf "%.${digits}f", $diff;

    my @parts;
    $days and push @parts, sprintf "%d day%s",   $days, $days == 1 ? "" : 's';
    $hour and push @parts, sprintf "%d hour%s",  $hour, $hour == 1 ? "" : 's';
    $mins and push @parts, sprintf "%d minute%s",$mins, $mins == 1 ? "" : 's';
    $diff && !$days && !$hour and push @parts, "$diff seconds";

    return join " ", @parts;
}

=head2 $record->as_hashref([$is_full])

Returns a HashRef with the inflated columns.

=head3 Parameters

Positional:

=over

=item 1. C<'full'>

If the word C<full> is passed as the first argument the related
C<configs> are also included in the resulting HashRef.

=back

=cut

sub as_hashref {
    my $self = shift;
    my ($is_full) = @_;
    $is_full //= '';

    my $record = { $self->get_inflated_columns };
    $record->{smoke_date} = $self->smoke_date->rfc3339 if $self->smoke_date;

    if ($is_full eq 'full') {
        $record->{configs} = [ map { $_->as_hashref($is_full) } $self->configs ];
    }

    return $record;
}

1;
